File: sdk/go/template.go
package liteboxd

import (
	"context"
	"strconv"
)

// TemplateService handles template operations.
type TemplateService struct {
	client *Client
}

// Create creates a new template.
func (t *TemplateService) Create(ctx context.Context, req *CreateTemplateRequest) (*Template, error) {
	var result Template
	err := t.client.doJSON(ctx, "POST", t.client.buildPath("templates"), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// List retrieves templates with optional filtering.
func (t *TemplateService) List(ctx context.Context, opts *TemplateListOptions) (*TemplateListResponse, error) {
	queryParams := make(map[string]string)
	if opts != nil {
		if opts.Tag != "" {
			queryParams["tag"] = opts.Tag
		}
		if opts.Search != "" {
			queryParams["search"] = opts.Search
		}
		if opts.Page > 0 {
			queryParams["page"] = strconv.Itoa(opts.Page)
		}
		if opts.PageSize > 0 {
			queryParams["pageSize"] = strconv.Itoa(opts.PageSize)
		}
	}
	var result TemplateListResponse
	err := t.client.doJSON(ctx, "GET", t.client.buildPath("templates"), nil, &result, queryParams)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Get retrieves a specific template by name.
func (t *TemplateService) Get(ctx context.Context, name string) (*Template, error) {
	var result Template
	err := t.client.doJSON(ctx, "GET", t.client.buildPath("templates", name), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Update updates a template (creates new version).
func (t *TemplateService) Update(ctx context.Context, name string, req *UpdateTemplateRequest) (*Template, error) {
	var result Template
	err := t.client.doJSON(ctx, "PUT", t.client.buildPath("templates", name), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Delete removes a template.
func (t *TemplateService) Delete(ctx context.Context, name string) error {
	return t.client.doEmptyResponse(ctx, "DELETE", t.client.buildPath("templates", name), nil, nil)
}

// ListVersions retrieves all versions of a template.
func (t *TemplateService) ListVersions(ctx context.Context, name string) (*VersionListResponse, error) {
	var result VersionListResponse
	err := t.client.doJSON(ctx, "GET", t.client.buildPath("templates", name, "versions"), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// GetVersion retrieves a specific version of a template.
func (t *TemplateService) GetVersion(ctx context.Context, name string, version int) (*TemplateVersion, error) {
	var result TemplateVersion
	err := t.client.doJSON(ctx, "GET", t.client.buildPath("templates", name, "versions", strconv.Itoa(version)), nil, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// Rollback rolls back a template to a previous version.
func (t *TemplateService) Rollback(ctx context.Context, name string, targetVersion int, changelog string) (*RollbackResponse, error) {
	req := &RollbackRequest{
		TargetVersion: targetVersion,
		Changelog:     changelog,
	}
	var result RollbackResponse
	err := t.client.doJSON(ctx, "POST", t.client.buildPath("templates", name, "rollback"), req, &result, nil)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// ExportYAML exports a template to YAML format.
func (t *TemplateService) ExportYAML(ctx context.Context, name string, version int) ([]byte, error) {
	queryParams := make(map[string]string)
	if version > 0 {
		queryParams["version"] = strconv.Itoa(version)
	}
	return t.client.doText(ctx, "GET", t.client.buildPath("templates", name, "export"), queryParams)
}

File: sdk/python/src/liteboxd/services/template.py
"""Template service for LiteBoxd SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models import (
    CreateTemplateRequest,
    RollbackResult,
    Template,
    TemplateListResult,
    TemplateVersion,
    UpdateTemplateRequest,
    VersionListResult,
    to_api_dict,
)

if TYPE_CHECKING:
    from ..client import Client


class TemplateService:
    """Template operations service."""

    def __init__(self, client: "Client") -> None:
        """Initialize template service."""
        self._client = client

    def create(self, request: CreateTemplateRequest) -> Template:
        """
        Create a new template.

        Args:
            request: Template creation request

        Returns:
            Created template object

        Raises:
            ConflictError: Template already exists
            BadRequestError: Invalid request parameters
        """
        response = self._client._post("templates", json=to_api_dict(request))
        return Template.model_validate(response.json())

    def list(
        self,
        *,
        tag: str | None = None,
        search: str | None = None,
        page: int = 1,
        page_size: int = 20,
    ) -> TemplateListResult:
        """
        List templates with optional filtering.

        Args:
            tag: Filter by tag
            search: Search in name, displayName, description
            page: Page number (default: 1)
            page_size: Items per page (default: 20, max: 100)

        Returns:
            Template list result with pagination info
        """
        params: dict[str, Any] = {
            "page": page,
            "pageSize": page_size,
        }
        if tag:
            params["tag"] = tag
        if search:
            params["search"] = search

        response = self._client._get("templates", params=params)
        return TemplateListResult.model_validate(response.json())

    def get(self, name: str) -> Template:
        """
        Get a specific template by name.

        Args:
            name: Template name

        Returns:
            Template object

        Raises:
            NotFoundError: Template not found
        """
        response = self._client._get(f"templates/{name}")
        return Template.model_validate(response.json())

    def update(self, name: str, request: UpdateTemplateRequest) -> Template:
        """
        Update a template (creates a new version).

        Args:
            name: Template name
            request: Update request

        Returns:
            Updated template object

        Raises:
            NotFoundError: Template not found
        """
        response = self._client._put(f"templates/{name}", json=to_api_dict(request))
        return Template.model_validate(response.json())

    def delete(self, name: str) -> None:
        """
        Delete a template.

        Args:
            name: Template name

        Raises:
            NotFoundError: Template not found
        """
        self._client._delete(f"templates/{name}")

    def list_versions(self, name: str) -> VersionListResult:
        """
        List all versions of a template.

        Args:
            name: Template name

        Returns:
            Version list result

        Raises:
            NotFoundError: Template not found
        """
        response = self._client._get(f"templates/{name}/versions")
        return VersionListResult.model_validate(response.json())

    def get_version(self, name: str, version: int) -> TemplateVersion:
        """
        Get a specific version of a template.

        Args:
            name: Template name
            version: Version number

        Returns:
            Template version object

        Raises:
            NotFoundError: Template or version not found
        """
        response = self._client._get(f"templates/{name}/versions/{version}")
        return TemplateVersion.model_validate(response.json())

    def rollback(
        self,
        name: str,
        target_version: int,
        *,
        changelog: str | None = None,
    ) -> RollbackResult:
        """
        Rollback template to a specific version.

        Args:
            name: Template name
            target_version: Version to rollback to
            changelog: Optional changelog message

        Returns:
            Rollback result

        Raises:
            NotFoundError: Template or version not found
        """
        payload: dict[str, Any] = {
            "targetVersion": target_version,
        }
        if changelog:
            payload["changelog"] = changelog

        response = self._client._post(f"templates/{name}/rollback", json=payload)
        return RollbackResult.model_validate(response.json())

    def export_yaml(
        self,
        name: str,
        *,
        version: int | None = None,
    ) -> bytes:
        """
        Export a template as YAML.

        Args:
            name: Template name
            version: Specific version (optional, defaults to latest)

        Returns:
            YAML content as bytes

        Raises:
            NotFoundError: Template not found
        """
        params: dict[str, Any] = {}
        if version is not None:
            params["version"] = version

        response = self._client._get(f"templates/{name}/export", params=params)
        return response.content
